src/hotspot/share/gc/shenandoah/heuristics/shenandoahAdaptiveHeuristics.cpp

```diff
@@ -145,24 +269,119 @@ bool ShenandoahAdaptiveHeuristics::should_start_gc() const {
+      break;
+    default:
+      ShouldNotReachHere();
+  }
+}
+
+void ShenandoahAdaptiveHeuristics::adjust_margin_of_error(double amount) {
+  _margin_of_error_sd = saturate(_margin_of_error_sd + amount, MINIMUM_CONFIDENCE, MAXIMUM_CONFIDENCE);
+  log_debug(gc, ergo)("Margin of error now %.2f", _margin_of_error_sd);
+}
+
+void ShenandoahAdaptiveHeuristics::adjust_spike_threshold(double amount) {
+  _spike_threshold_sd = saturate(_spike_threshold_sd - amount, MINIMUM_CONFIDENCE, MAXIMUM_CONFIDENCE);
+  log_debug(gc, ergo)("Spike threshold now: %.2f", _spike_threshold_sd);
+}
+
+ShenandoahAllocationRate::ShenandoahAllocationRate(ShenandoahAdaptiveHeuristics *heuristics) :
+  _heuristics(heuristics),
+  _last_sample_time(os::javaTimeNanos()),
+  _last_sample_value(0),
+  _interval_ns(NANOUNITS / ShenandoahAdaptiveSampleFrequencyHz),
+  _rate(ShenandoahAdaptiveSampleSizeSeconds * ShenandoahAdaptiveSampleFrequencyHz, ShenandoahAdaptiveDecayFactor),
+  _rate_avg(ShenandoahAdaptiveSampleSizeSeconds * ShenandoahAdaptiveSampleFrequencyHz, ShenandoahAdaptiveDecayFactor) {
+}
+
+void ShenandoahAllocationRate::sample(size_t bytes_allocated_since_gc_start) {
+  jlong now = os::javaTimeNanos();
+  if (now - _last_sample_time > _interval_ns) {
+    if (bytes_allocated_since_gc_start > _last_sample_value) {
+      size_t allocation_delta = bytes_allocated_since_gc_start - _last_sample_value;
+      size_t time_delta_ns = now - _last_sample_time;
+      double alloc_bytes_per_second = ((double) allocation_delta * NANOUNITS) / time_delta_ns;
+
+      _rate.add(alloc_bytes_per_second);
+      _rate_avg.add(_rate.avg());
+    }
+
+    _last_sample_time = now;
+    _last_sample_value = bytes_allocated_since_gc_start;
+  }
+}
+
+double ShenandoahAllocationRate::upper_bound(double standard_deviations) const {
+  // Here we are using the standard deviation of the computed running
+  // average, rather than the standard deviation of the samples that went
+  // into the moving average. This is a much more stable value and is tied
+  // to the actual statistic in use (moving average over samples of averages).
+  return _rate.davg() + (standard_deviations * _rate_avg.dsd());
+}
+
+void ShenandoahAllocationRate::allocation_counter_reset() {
+  _last_sample_time = os::javaTimeNanos();
+  _last_sample_value = 0;
+}
+
+bool ShenandoahAllocationRate::is_spiking(double instantaneous_rate) const {
+  double standard_deviation = _rate.sd();
+  if (standard_deviation > 0) {
+    // There is a small chance that that rate has already been sampled, but it
+    // seems not to matter in practice.
+    double z_score = (instantaneous_rate - _rate.avg()) / standard_deviation;
+    if (z_score > _heuristics->_spike_threshold_sd) {
+      return true;
+    }
+  }
+  return false;
+}
+
+double ShenandoahAllocationRate::instantaneous_rate(size_t bytes_allocated_since_gc_start) const {
+  size_t allocation_delta = bytes_allocated_since_gc_start - _last_sample_value;
```

Comment:
`allocation_delta` underflows here in my benchmarks, causing back-to-back "Instantaneous rate" triggers. Note that `sample()` method seems to check for it.

Reply:
Yep, that's because I fumbled the `allocation_counter_reset` call.